### ISSUE TITLE
Table of Contents does not include headings not added as heading blocks

### ISSUE THREAD

**Issue by bethdeconinck** [WordPress/gutenberg#5037]
## Issue Overview
I don't think I've seen this in any existing issues. This is a relatively small issue, but definitely something that shouldn't fall through the cracks.

There are three supported ways to add headings to a post:
- By adding a **Heading** block
- By adding a **Classic** editor block which contains headings in its contents
- By adding a **Custom HTML** block and hand-coding an `<h2>`, `<h3>`, etc.

It's easy to imagine plugins adding their own blocks that include headings in their markup (including, hopefully, a heading-level selector to choose the right type of heading).

The Table of Contents/Document Outline generates an outline of the structure of the document, which is enormously useful to help content creators build accessible content. However, the ToC only contains headings added as **Heading** blocks.

Not sure how the ToC is being generated, but it should give a complete picture of the heading structure of the post---any headings in the post markup, as end-users will experience it. Otherwise, content creators who come to rely on the ToC feature may unwittingly give their posts poor or confusing structure.

## Steps to Reproduce (for bugs)

1. Create a new post using Gutenberg.
2. Add a **Heading** block, H2: "Heading 2 in a Heading block"
3. Add a **Classic** editor block, H2: "Heading 2 in a Classic Editor block"
4. Add a **Custom HTML** block, and code a `<h2>Heading 2 in a Custom HTML block</h2>`.
5. Open the Table of Contents (right-hand sidebar) or the Document Outline (top left-hand corner of the editor, in the Content Structure button.

This is in Gutenberg v2.1.0, WP v4.9.4, Google Chrome v64.

## Expected Behavior

The Table of Contents should show the document title followed by three H2 headings.

## Current Behavior

The Table of Contents shows the document title followed by just one H2 heading, "Heading 2 in a Heading block."

## Screenshots / Video

![gutenberg toc - heading issue](https://user-images.githubusercontent.com/6844775/36171548-10e276ca-10d1-11e8-9c9d-5aaf8366592e.png)

**Comment by youknowriad** [comment]
The Table of contents looks specifically to the presence of some defined blocks. I'd consider this a minor enhancement, but having a generic API to add items to the table of contents seems legit.

**Comment by mrwweb** [comment]
@karmatosed @youknowriad It seems odd that this issue is closed, since it's still an problem and feels very much like a bug (the title could be rewritten to read "Document Outline only includes heading blocks when it should include all headings"). This is related to #33079 where the outline doesn't see Reusable Blocks. I'm currently working on an accordion block where I would expect the headings (output into the post content via the `save` function) to be included in the Document Outline.

It seems that this getting closed has meant it won't get any attention. Can it be reopened since it's still an issue?

**Comment by youknowriad** [comment]
@mrwweb Right now the headings/blocks highlighted are hardcoded, if we want third-party (and reusable blocks) support, we'd need a dedicated Block API potentially that takes all of the use-cases into consideration.

**Comment by mrwweb** [comment]
@youknowriad Thanks for reopening! I hope this can get some eyes on it in the coming versions!

I wonder if it would be possible to make the outline by parsing the rendered editor contents? Between the Classic block, HTML block, Reusable Blocks, and 3rd-party blocks, it seems that making this change would likely make the outline more useful in a significant number of cases and also help highlight heading outline problems more frequently.

**Comment by cpescha** [comment]
Yes, definitely it would be very helpful and important for SEO optimisation reasons having a complete heading structure of the post. Would be great to see it implemented. 
Thanks

**Comment by leogitdev** [comment]
I'm writing a custom heading block that can have h2~h6 tags and I think it's unfortunate that there's no way for them to show in the document outline in the sidebar.

**Comment by laubsterboy** [comment]
Is there any progress on this? Is there anything I can do to help this get moved along? I have multiple custom blocks that have heading tags in them but don't use InnerBlocks because I need to enforce strict styles and can't allow the user to add more than one heading, but then my heading doesn't get incorporated into the Table of Contents so the hierarchy appears to be broken but it's not (or could appear to be good but be broken).

**Comment by Mamaduka** [comment]
Closing as duplicate in favor of #71699 to avoid confusion with ToC block, where a similar request is tracked separately (#54646).

The issue refers to the old ToC, now rendered as Document Outline.
